modules/broker-core/src/main/java/org/wso2/broker/core/MessagingEngine.java

```diff
@@ -183,28 +185,19 @@ void publish(Message message) throws BrokerException {
                 BindingSet bindingSet = exchange.getBindingsForRoute(routingKey);
 
                 if (bindingSet.isEmpty()) {
-                    LOGGER.info("Dropping message since no queues found for routing key " + routingKey + " in "
-                                        + exchange);
+                    LOGGER.info("Dropping message since no queues found for routing key {} in {}",
+                                routingKey, exchange);
                     message.release();
                     MessageTracer.trace(message, MessageTracer.NO_ROUTES);
                 } else {
                     try {
-                        sharedMessageStore.add(message);
-                        Set<String> uniqueQueues = new HashSet<>();
-                        for (Binding binding : bindingSet.getUnfilteredBindings()) {
-                            uniqueQueues.add(binding.getQueue().getName());
-                        }
-
-                        for (Binding binding : bindingSet.getFilteredBindings()) {
-                            if (binding.getFilterExpression().evaluate(metadata)) {
-                                uniqueQueues.add(binding.getQueue().getName());
-                            }
-                        }
+                        sharedMessageStore.addShallowCopy(message);
```

Comment:
Do we need to do a shallow copy here? It seems we haven't used it anywhere else.

Reply:
We need to shallow copy since message storage is handled with a separate copy of the message